File: src/mk2/PythonPackage/src/tamaf/sequentialbehavior.py
from .behavior import Behavior

class SequentialBehavior(Behavior):
    def __init__(self, agent):
        super().__init__(agent)
        self.subBehaviors: list[Behavior] = []
        self.currentIndex: int = 0
        self.sharedData: dict = {}

    def AddSubBehavior(self, behavior: Behavior):
        behavior.agent = self.agent
        self.subBehaviors.append(behavior)

    def OnStart(self):
        if self.currentIndex < len(self.subBehaviors):
            self.subBehaviors[self.currentIndex].OnStart()
    
    def Action(self):
        if self.currentIndex < len(self.subBehaviors):
            self.subBehaviors[self.currentIndex].Execute()
            if self.subBehaviors[self.currentIndex].IsBlocked():
                self.Block()

    def Done(self):
        if self.currentIndex >= len(self.subBehaviors):
            return True
            
        if not self.subBehaviors[self.currentIndex].Done():
            return False
        else:
            self.subBehaviors[self.currentIndex].OnEnd()
            self.currentIndex += 1
            if self.currentIndex >= len(self.subBehaviors):
                return True
            else:
                self.isStarted = False
                return False

    def OnEnd(self):
        self.sharedData = {}
        self.currentIndex = 0
        self.isStarted = False
        return 0

File: src/mk3/JavaPackage/src/main/java/tamaf/SequentialBehavior.java
package tamaf;

import java.util.ArrayList;
import java.util.List;

public class SequentialBehavior extends Behavior {
    private List<Behavior> subBehaviors = new ArrayList<>();
    private int currentIndex = 0;
    protected java.util.Map<String, Object> sharedData = new java.util.HashMap<>();

    public SequentialBehavior() {
        super();
    }
    
    public SequentialBehavior(Agent agent) {
        super(agent);
    }

    public void addSubBehavior(Behavior b) {
        b.setAgent(this.agent);
        subBehaviors.add(b);
    }

    @Override
    public void onStart() {
        if (!subBehaviors.isEmpty()) {
            subBehaviors.get(currentIndex).onStart();
        }
    }

    @Override
    public void action() {
        if (currentIndex < subBehaviors.size()) {
            Behavior current = subBehaviors.get(currentIndex);
            current.execute();
            if (current.isBlocked()) {
                this.block();
            }
        }
    }

    @Override
    public boolean done() {
        if (subBehaviors.isEmpty()) return true;
        if (currentIndex >= subBehaviors.size()) return true;

        Behavior current = subBehaviors.get(currentIndex);
        if (current.done()) {
            current.onEnd();
            currentIndex++;
            if (currentIndex >= subBehaviors.size()) {
                return true;
            } else {
                setStarted(false); // Reset to trigger onStart of next
                return false;
            }
        }
        return false;
    }

    @Override
    public int onEnd() {
        this.sharedData.clear();
        this.currentIndex = 0;
        this.setStarted(false);
        return 0;
    }
}
